src/gz/bufread.rs

```diff
@@ -338,34 +476,45 @@ impl<'a, T: Read> Read for Buffer<'a, T> {
     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
         if self.buf_cur == self.buf_max {
             let len = self.reader.read(buf)?;
-            self.buf.extend_from_slice(&buf[..len]);
+            self.part.buf.extend_from_slice(&buf[..len]);
             Ok(len)
         } else {
-            let len = (&self.buf[self.buf_cur..self.buf_max]).read(buf)?;
+            let len = (&self.part.buf[self.buf_cur..self.buf_max]).read(buf)?;
             self.buf_cur += len;
             Ok(len)
         }
     }
 }
 
+impl<'a, T> Buffer<'a, T> where T: std::io::Read {
+    // If we manage to read all the bytes, we reset the buffer
+    fn read_once(&mut self, buf: &mut [u8]) -> io::Result<usize> {
```

Comment:
I did not see any good fit in neither `BufRead` nor `Read` for this...

What I want is
`read_exact` but buffer it there is not enough data, and clear the buffer if it succeeded

Is there some more idiomatic rust way ?